### ISSUE TITLE
Maintainable Architecture

### ISSUE THREAD

**Issue by adrian-moisa** [visual-space/visual-editor#1]
It will be a while until visual editor is all packed up and ready to serve duty.  At best at least 1 month until we get testing and docs such that we can provide a better service. (I mean it's currently usable as it is a direct clone of Flutter Quill). I've been already going trough the steps to setup the repo.  There's a lot to review. Today I started work on the cleanup effort. There's a lot to unpack here. My first goal, not have files with over 300 LOC. And to split in modules. Plus rename / document where necessary to make it easy to follow the code flow. 

Join on [discord](https://discord.gg/XpGygmXde4) to get advice and help or follow us on [YouTube Visual Coding](https://www.youtube.com/channel/UC2-5lfNbbErIds0Iuai8yfA) to learn more about the architecture of Visual Editor and other Flutter apps.

**Comment by adrian-moisa** [comment]
Making good progress on the refactoring 
![unknown](https://user-images.githubusercontent.com/11160141/168167792-44174f4d-9146-4b8d-a11f-301fbe4f63e8.png)

just a small hint how bad it is/was... 
the entry file had 2000 LOC and 8 freaking classes....
editor.dart is now spread in several files. I'll have more work to do in order to properly document them.

<img width="344" alt="25 editor split" src="https://user-images.githubusercontent.com/11160141/168167854-388d3f56-0c6c-4471-b18e-6f2ac0d21b6b.png">

**Comment by adrian-moisa** [comment]
I've completed the task of splitting everything into modules. Here's a breakdown of the numbers so far.
<img width="713" alt="41a Before" src="https://user-images.githubusercontent.com/11160141/168468778-bff3cbef-634d-4e5e-91e1-cff18fcd22c0.png">
<img width="714" alt="41b after" src="https://user-images.githubusercontent.com/11160141/168468779-8e696334-9011-482c-8987-afdd2c47e1ee.png">
<img width="385" alt="41c" src="https://user-images.githubusercontent.com/11160141/168468780-bd815710-f4b1-4ef0-bda2-c92f9c1a63e3.png">

**Comment by apalala-dev** [comment]
Great work already 💪

I have a few other things in mind that might improve the usage of this package as well.

### Tuple package dependency

Atm, there is a dependency to the tuple package which I think should be either removed or should be exposed so visual-editor users won't have to import the tuple package as well.
I add to import the package manually to edit the lineSpacing for instance:

``` dart
quill.VisualEditor(
                customStyles: quill.DefaultStyles(
                    paragraph: quill.DefaultTextBlockStyle(
                  Theme.of(context).textTheme.bodyText2!,
                  const Tuple2(8, 0),
                  const Tuple2(0, 0),
                  null,
                )),
// ....
)
```

This seems to have been used a lot in the package.
However, having a Tuple instead of a concrete class makes the implementation quite obscure. Here for instance, I don't see why the lineSpacing is not a simple double instead of a Tuple.

### Name conflicts

Some names in the package are very generic and might insert name conflicts when using VisualEditor.
For instance, there is the `Text` class which extends `Leaf`. There is already a very common `Text` class in Flutter which might make people have conflicts.

It is more a problem for new users since one just needs to use an alias for the import as a workaround, but I guess a TextLeaf would have been easier.

**Comment by adrian-moisa** [comment]
@apalala-dev Fully agree with both comments. You are very right to say that Tuple obscures meaning. I'll add this to my refactorings list. As for the Text name conflict I'm already aware and working to get rid of it. I've encountered it myself many times. High prio. Once again. Thank you for the thorough feedback! It's much appreciated.

**Comment by adrian-moisa** [comment]
The RawEditor is getting in a far better shape. As you can see, everything is now split in nicely named services. This will pay big dividends in the long run. 
<img width="1145" alt="16 new services" src="https://user-images.githubusercontent.com/11160141/173159887-4f1ab15d-a26c-4f3c-a4f8-e98509c5c7bf.png">

**Comment by adrian-moisa** [comment]
**Major Milestone - Editor and RawEditor classes are finally merged in one single widget 🥳** 

After several weeks of refactoring I've finally been able to puzzle together the role of each layer in the hierarchy of Quill. Eventually I figured out that there is no fundamental distinction between QuillEditor (aka VisualEditor) and RawEditor. For a very long time I was deeply confused what is the meaning of Editor vs RawEditor. This was one of the most confusing parts of the codebase. Now no longer, they are fused in one single widget and the build() method is now crystal clear as you can see bellow.

Still a lot more work ahead. However this marks a major point in the refactoring process. We are over the (complexity) hill. From here on it should be a lot easier to continue the refactoring effort. Also a lot easier to provide documentation since now, it makes a lot more sense. I'm also very happy to report that we don't have too many regressions so far out of this refactoring. I did my best to constantly check that we are retaining all features.

By now I have content and scripts for 6 episodes (and 40 more planned ahead). I will start allocating time for the voice overs to start releasing them.

<img width="1146" alt="11c" src="https://user-images.githubusercontent.com/11160141/173927246-71089c89-0d89-425a-97f4-e0d36b1f0666.png">

**Comment by adrian-moisa** [comment]
I've reached a point where I consider the main project structure clean enough to be easily maintainable (at least by myself). I still have a lot of work to do in documentation and testing department, but this will come in time as I work on other features. From here on I continue work on the code base in a need to have order. Further code improvements will be done as part of ongoing features tickets or bugs. So this one will be closed.

The current state of the editor is unstable. It has a few new issues stemming from the refactoring effort. So until it's not clean enough I don't plan to release it on pub.dev. Until then it can be imported straight from github. 

    visual_editor:
      git: https://github.com/visual-space/visual-editor.git
    
By now I've collected materials for the first 6 episodes. I'll be working in the background to get them edited and uploaded on our [YT channel](https://www.youtube.com/channel/UC2-5lfNbbErIds0Iuai8yfA).

<img width="940" alt="Screenshot 2022-06-25 at 22 39 24" src="https://user-images.githubusercontent.com/11160141/175789850-3740d941-a2b2-42b4-8916-a001e101cc70.png">
